accepted/wasmtime-one-dot-oh.md

```diff
@@ -0,0 +1,558 @@
+changes makes it feel harder to use Wasmtime since it can feel more difficult to
+keep up with the latest and receive bug fixes and new features. Wasmtime's
+stability story, consequently, will not be on the end of the spectrum where
+breaking changes are landed willy-nilly. This means that there's a tension set
+up between developers of Wasmtime and users of Wasmtime, but this is to be
+expected from the stable release of a project.
+
+On a spectrum of "Wasmtime never breaks any API ever" to "Wasmtime breaks all
+its APIs every release" Wasmtime will reside towards the first half here. In
+other words Wasmtime developers will be expected to rarely and deliberately
+break existing APIs, but will also be allowed to break APIs as necessary. The
+major goal here is "hassle-free upgrades" inspired by the Rust release process
+where API stability in the absolute literal sense is detrimental to the
+development of Wasmtime, but deliberate effort and work is done to ensure
+upgrades are as seamless as possible.
+
+Users should expect that they will regularly need to increase the major version
+of Wasmtime to upgrade Wasmtime usage in their projects. Furthermore users
+should expect that between upgrades they may need to tweak some APIs and update
+them to continue to use the latest-and-greatest Wasmtime. At the same time
+though users should expect that they will not be blindsided by major
+refactorings. For example changes like [Wasmtime's new API][new-api] will still
+go through the RFC process and will be deliberately scheduled and widely
+advertised before they're released. Note also that changes of that magnitude are
+expected to be quite rare.
+
+[new-api]: https://github.com/bytecodealliance/rfcs/blob/main/accepted/new-api.md
+
+## What is being stabilized?
+
+Wasmtime is a relatively large project at this point with lots of components.
+Additionally not everything lives in the `wasmtime` repository itself but there
+are separate dependencies such as `wasm-tools`, `wasmtime-*` embeddings,
+`witx-bindgen`, etc. This proposal does not attempt to blindly label everything
+with 1.0 and ship it, but rather discrete "tiers" of support are defined for
+what it means to use a component of the Wasmtime project.
+
+At a base level anything shipped in Wasmtime is guaranteed to clear a minimum
+threshold of quality. All dependencies, even the transitive ones, used by
+Wasmtime are guaranteed to be "production quality" for the surface area that
+Wasmtime uses. On the other hand, though, not all of Wasmtime's dependencies
+and components will meet the same level of API stability expected of a stable
+project. For example Cranelift is not an API-stable component that will be
+released with Wasmtime. For now Cranelift will likely have the same release
+cadence of Wasmtime for ease of infrastructure, but this may change in the
+future as Cranelift develops its own API and release process. Additionally,
```

Comment:
It might be an open question at the moment, but still curious: until there is such a separate Cranelift release process, does it mean that the Cranelift shipped with Wasmtime LTS is actually in LTS mode itself? It might have different requirements in terms of backports for bugfixes and sec fixes (e.g. some of these might involve only Cranelift and nothing in Wasmtime, and conversely), so this really calls for a different release process as you suggest. Still, it would be interesting to get a clear stand on where Cranelift is at the moment in terms of support and will be once this RFC gets merged.

Reply:
In my view at least, it makes sense for us to uphold the same LTS promises with Cranelift as with Wasmtime with regard to bug-fixes, because a security bug in Cranelift *is* a security bug in Wasmtime from the point of view of the consumer. So, yes, I would expect that we backport any correctness or security fix to codegen.

(Also FWIW, it's hard for me to imagine at the moment what would cause Cranelift to diverge to another release process, as riding along with Wasmtime gives all of the needed framework for managing stable branches, security patches, etc. -- at most we might introduce a stronger API-stability guarantee on `cranelift-codegen` and friends -- but for completeness, if Cranelift does ever diverge, I would expect it to keep a very similar LTS and backports policy.)

Reply:
My view on this is that as part of Wasmtime's 1.0 stability promise we're promising quality in our transitive dependencies, including cranelift, so any cranelift bugs affecting Wasmtime is something we have to take care of in Wasmtime (patch releases, etc).

If there's a bug in Cranelift, though (security or not), which we probably know cannot affect Wasmtime (e.g. in an architecture not supported or something like that) I think the process is a bit more ambiguous. Most of this RFC is about LTS/point releases for Wasmtime, and I don't think it would apply in the case that the bug from Cranelift isn't exposed through Wasmtime's usage. 

Not to say that point releases *won't* happen, just that I don't think it would always necessitate a Wasmtime point release as well. Ideally the release process for these two would be diverged so that Cranelift could be released without the other (and in theory Wasmtime as well if Cranelift didn't have changes, but that seems unlikely), but that may take some more infrastructure.

Reply:
> If there's a bug in Cranelift, though (security or not), which we probably know cannot affect Wasmtime (e.g. in an architecture not supported or something like that) I think the process is a bit more ambiguous. Most of this RFC is about LTS/point releases for Wasmtime, and I don't think it would apply in the case that the bug from Cranelift isn't exposed through Wasmtime's usage.

I think this is the key point. The way Cranelift is tested right now in particular doesn't really allow us to make strong claims about its stability and quality properties when used to compile non-Wasm content, and perhaps even when used in other embeddings in general.

I think that will change, but this RFC is about Wasmtime, and in that context it makes sense to treat Cranelift purely as a dependency *to Wasmtime*, and look at its stability and quality properties through that lens.

>  it's hard for me to imagine at the moment what would cause Cranelift to diverge to another release process

I also agree with this, though! When Cranelift gains a stability story of its own, that explicitly encompasses other uses, I don't think that'd be a reason for it to also gain a completely separate release process. It seems much better to broaden the scope of guarantees we give for it by introducing the required testing strategies, etc, but still let it benefit from being included in the same release train schedule, etc.